File: 5. Longest Palindromic Substring.java
class Solution {
    public String longestPalindrome(String t) {
        int n = t.length(), m = 2*n +1,mx =0;
        String s = "#";
        for(int i=0;i<n;i++){
            s += t.charAt(i) + "#";
        }
        int []dp = new int[m];
        int center = 0, right = 0;
        for(int i =0;i<m;i++){
            if(i < right) dp[i] = Math.min(right - i, dp[2*center - i]);
            while(i > dp[i] && i + dp[i] + 1 < m && s.charAt(i-dp[i] - 1) == s.charAt(i + dp[i] + 1)) dp[i]++;
            if(i + dp[i] > right) {
                center =i;
                right = i + dp[i];
            }
            if(dp[i] > dp[mx]) mx = i; 
        }
        return s.substring(mx - dp[mx] , mx + dp[mx] + 1).replaceAll("#","");
    }
};

File: 5. Longest Palindromic Substring.cpp
/*
Problem credits: https://leetcode.com/problems/longest-palindromic-substring/

Given a string s, return the longest palindromic substring in s.

Solution: Manacher algorithm

*/
class Solution {
public:
    string longestPalindrome(string s) {
        s = "^" + regex_replace(s, regex(""), "#") + "$";
        int n = s.size();
        string ans;
        vector<int>radius(n, 0);
        int l = 0, r = 1;
        for(int i =1; i < n-1; i++) {
            int mirror = (l + r) - i;
            radius[i] = min(r - i, radius[mirror]);
            while (s[i - radius[i]] == s[i+radius[i]]) {
                radius[i]++;
            }
            if (i + radius[i] > r) {
                l = i - radius[i];
                r = i + radius[i];
                int sz = 2* radius[i] - 1;
                if (ans.size() < sz) {
                    ans = s.substr(l+1, sz);
                }
            }
        }
        return regex_replace(ans, regex("#"), "");
    }
};